### PR TITLE
Add label-matcher support to Rules API

### LINKED ISSUE #9627: Rules API: support label-based matcher. [Feature Request]
## Proposal
**Use case. Why is this important?**
In thanos-sidecar mode, fanout prometheus rules api may response lots of rules. Could we support label based matcher in prometheus rules api, like series api `match[]` parameters . https://github.com/prometheus/prometheus/blob/main/docs/querying/api.md#finding-series-by-label-matchers . 😃 
relevant issue in thanos https://github.com/thanos-io/thanos/issues/4812

### PR DESCRIPTION
**Pull request by saswatamcode**
This PR adds support for label matchers as URL query params to the `api/v1/rules` endpoint of Prometheus API. This will not break existing functionality.

It filters both recording and alerting rules via their labels (both templated and non-templated).

Open question on whether active alerts should be handled differently?

A major use case for this is multi-tenant applications built on Prometheus which require a way to isolate tenant rules. Currently, one possible way of doing this is by filtering the API response. But, for endpoints like `api/v1/labels`, only a matcher needs to be added to URL params. So this PR aims to bring the same convenience to rules.

Fixes #9627.

### CODE REVIEW

**Comment by roidelapluie** [comment]
While it makes sense to filter active alerts, it does not make sense to filter alert and rules definition as the label value is a templated string. I am against adding this feature.

**Comment by roidelapluie** [comment]
I have also commented on the issue.

File: docs/querying/api.md
Comment (bwplotka) [inline docs/querying/api.md:641]: Let's start with one API first if we do this.

File: rules/manager.go
Comment (bwplotka) [inline rules/manager.go:326]: I guess we need special label matching for templates defined 🤔

File: web/api/v1/api.go
Comment (bwplotka) [inline web/api/v1/api.go:1048]: Let's kill it for now, to keep discussion simpler I guess

File: rules/manager.go
Comment (bwplotka) [reply rules/manager.go:326]: Or we keep it simple and update docs instead.

File: docs/querying/api.md
Comment (bwplotka) [inline docs/querying/api.md:578]:
```suggestion
- `match[]=<label_selector>`: Repeated label selector argument that selects rules which have configured labels that satisfy the selectors. Note that alerting rules it matches the top-level potentially templated labels, not the rendered labels of active alerts (instances). Optional.
```

File: web/api/v1/api_test.go
Comment (bwplotka) [inline web/api/v1/api_test.go:1576]: Since we have cases of templated labels, let's test them as well to show that we match those in non-rendered form.

**Review by bwplotka — COMMENTED** [review]
Thanks @saswatamcode for this PR and @roidelapluie for review.

Let's start with Rules API first as this is the API that matters the most.

And as per your point @roidelapluie  I see this being weird from the first glance, but what if we provide the most straightforward mechanism that will be expected by everyone. Rules API is all about loaded configured, so to me we should filter by non-rendered, configured labels. This is also how it is implemented right now in this PR. I proposed a doc entry. What's wrong with that @roidelapluie 🤔

**Comment by roidelapluie** [comment]
What is wrong is the semantic. I'd rather have match[] only match labels, and not templates in one API and label values in another API. I'd be okay if we would make this available in a different way, but *not* via a "promql-like" selector, because it does not make sense to apply regex on golang templates, it's mixing apples and oranges.

**Comment by saswatamcode** [comment]
As @bwplotka [suggested](https://github.com/prometheus/prometheus/pull/10194#discussion_r792577293) then, maybe we can keep it in the simplest form and match only non-templated labels in rules (with a change in docs to highlight this)? That way match[] doesn't behave differently from other APIs and we don't apply regex on templates. WDYT @roidelapluie? 🙂

**Comment by roidelapluie** [comment]
I trust Bartek's judgment, let's document this properly.

File: web/api/v1/api.go
Comment (saswatamcode) [reply web/api/v1/api.go]: Yes, have removed this!

File: docs/querying/api.md
Comment (saswatamcode) [reply docs/querying/api.md]: Updated docs with a slight change, we only match non-templated labels.

File: web/api/v1/api_test.go
Comment (saswatamcode) [reply web/api/v1/api_test.go:1599]: Added a test, but showcased as template label in matcher not returning anything!

File: rules/manager.go
Comment (saswatamcode) [reply rules/manager.go:328]: Updated docs instead of matching templated labels. We instead check if the label values are templated. If they are templated they are skipped while matching, We do this by parsing the label value as a template and check for the number of nodes. If there is only one node and the type is [`parser.NodeText`](https://pkg.go.dev/text/template/parse@go1.17.6#NodeType) then it is treated as a non-templated label. Should this be a regex-based approach instead? cc: @roidelapluie

**Comment by saswatamcode** [comment]
I have updated this PR with a doc change. Now only `api/v1/rules` support matchers and that too for non-templated label values only (approach for detecting templating is using parser node type, but can be changed to a regex-based one)! 🙂

**Comment by saswatamcode** [comment]
Is there anything else that seems to be missing or is blocking this PR, @roidelapluie? Would love to progress on this! 🙂

**Comment by qinxx108** [comment]
Regarding to only match non-templated labels in rules. From user usage perspective, i should be able to filter based on what i configured? For example as in the test cases, if my rule have a label 
`Name: "templatedlabel", Value: "{{ $externalURL }}"`
If i search for templatedlabel= {{ $externalURL }}, it should return to me the rules that has `templatedlabel= {{ $externalURL }} `

File: rules/manager.go
Comment (qinxx108) [inline rules/manager.go:1191]: From my perspective this should be OR following the other filter this api support in this PR: https://github.com/prometheus/prometheus/pull/12270/files

That is if there is a list of matcher to match, if any of them match the label, we should return rule.

Comment (saswatamcode) [reply rules/manager.go]: You're right. I changed the logic, to match any matcherset, instead of all. Also rebased the PR.

**Comment by saswatamcode** [comment]
Hmm, not sure why the test/build fails here...maybe due to stringlabels?

**Comment by alvinlin123** [comment]
> Regarding to only match non-templated labels in rules. From user usage perspective, i should be able to filter based on what i configured? For example as in the test cases, if my rule have a label `Name: "templatedlabel", Value: "{{ $externalURL }}"` If i search for templatedlabel= {{ $externalURL }}, it should return to me the rules that has `templatedlabel= {{ $externalURL }} `

I agree with this. While I understand @roidelapluie view about the semantic of `match` for templated string,  as @bwplotka said "...Rules API is all about loaded configured" so it makes perfect sense the filter parameter works against whatever is configured, which includes the templated string. As a system admin, there are legit use cases where I want to do something like "gimme all the rule definition that has `{{ $someTypo }}`" because I made a boo boo and I want to production impact analysis and fix the issue. 

For 
> maybe we can keep it in the simplest form and match only non-templated labels in rules
 
I am against this because it's not intuitive. Sure we can document it, but I think it's simpler for the the suer to understand "the filter works against template string" than "if the value is a template string then ... otherwise ...". Math only non-templated labels seems like an inconsistent behaviour to me.

**Comment by roidelapluie** [comment]
Okay.

**Comment by alvinlin123** [comment]
> Okay

Love the conciseness :D

**Comment by bboreham** [comment]
> Hmm, not sure why the test/build fails here...maybe due to stringlabels?

The [error log](https://github.com/prometheus/prometheus/actions/runs/6401993179/job/17377888011?pr=10194) says:
```
rules/manager.go:467:23: cannot range over l (variable of type labels.Labels)
rules/manager.go:485:24: cannot range over l (variable of type labels.Labels)
```

It was never a great idea to assume `labels.Labels` was a slice, and #10991 made this concrete.

You could use code like this: https://github.com/prometheus/prometheus/blob/617bee60f19bfae850ffc6386e4be7637a0cb631/model/labels/test_utils.go#L34-L41

**Comment by saswatamcode** [comment]
Thanks @bboreham!

@alvinlin123 / @roidelapluie let me know, if we should remove the templated labels check here then.

**Comment by roidelapluie** [comment]
Yes please

**Comment by saswatamcode** [comment]
Done!

File: docs/querying/api.md
Comment (qinxx108) [inline docs/querying/api.md:682]: should we adjust the doc as well?

Comment (saswatamcode) [reply docs/querying/api.md]: Done in https://github.com/prometheus/prometheus/pull/10194/commits/711dbc1041c53e316679a3727dccabaef5474163. Updated and changed to make it consistent with rest of the params.

Comment (qinxx108) [reply docs/querying/api.md]: ty!

File: rules/manager.go
Comment (qinxx108) [reply rules/manager.go:467]: fixed

File: rules/manager.go
Comment (qinxx108) [reply rules/manager.go:482]: the other query parameters for example 

https://github.com/prometheus/prometheus/pull/12270/files

**Comment by qinxx108** [comment]
@roidelapluie could you please take a look at this pr again? Thanks

File: rules/manager.go
Comment (bboreham) [reply rules/manager.go:479]: My expectation is that all the Matchers in each `[]*labels.Matcher` will be ANDed, and then the set of those will be `ORed`.  I think this is what we do elsewhere.  
It could be made clearer in comments.

File: docs/querying/api.md
Comment (bboreham) [inline docs/querying/api.md:682]: I think this text is clearer:
```suggestion
- `match[]=<label_selector>`: only return rules that have configured labels that satisfy all the label selectors. If the parameter is repeated, rules that match any of the sets of label selectors are returned. Note that matching is on the labels in the definition of each rule, not on the values after template expansion (for alerting rules). Optional.
```

File: web/api/v1/api_test.go
Comment (bboreham) [inline web/api/v1/api_test.go:215]: Since it doesn't use the parameter.
```suggestion
func (m rulesRetrieverMock) AlertingRules(_ ...[]*labels.Matcher) []*rules.AlertingRule {
```

(even better if the test did use the parameter)

File: web/api/v1/api_test.go
Comment (bboreham) [inline web/api/v1/api_test.go:2264]: Suggest comment that this case is testing two conditions 'OR'ed.

File: web/api/v1/errors_test.go
Comment (bboreham) [inline web/api/v1/errors_test.go:256]:
```suggestion
func (DummyRulesRetriever) AlertingRules(_ ...[]*labels.Matcher) []*rules.AlertingRule {
```

**Review by bboreham — COMMENTED** [review]
Code looks broadly fine; I made some small notes in places.

> A major use case for this is multi-tenant applications built on Prometheus
Having worked on Cortex and Mimir, I don't understand this part. Both of them take a tenant ID in the request, and return rules only for that tenant.
Can you plese update the description to match what you want to get merged?

**Comment by alvinlin123** [comment]
> A major use case for this is multi-tenant applications built on Prometheus
> Having worked on Cortex and Mimir, I don't understand this part. Both of them take a tenant ID in the request, and return rules only for that tenant.
> Can you plese update the description to match what you want to get merged?

Hi @bboreham!  Long time no "see"! 

Yea initially I found this use case head scratching too. But I think the original author is implementing some kind of "poor man's multi-tenancy" on top of just Prometheus using just labelling convention for metrics and rules.

**Comment by saswatamcode** [comment]
> I think the original author is implementing some kind of "poor man's multi-tenancy" on top of just Prometheus using just labelling convention for metrics and rules

That's correct, this is what the PR aims to do! We already kind of have this in Thanos, and can just use a proxy in front to ensure rules with a particular tenant label get returned, without having to maintain the notion of tenancy internally. 

This isn't a new pattern I guess and can extend beyond tenancy. Even projects like prom-label-proxy need to modify rules response and filter currently, when just setting a `match[]` param would be more convenient.

> Can you please update the description to match what you want to get merged?

The original description would still be valid I think, but I'm happy to change it to something more suitable. Open to any suggestions cc: @qinxx108  🙂

**Comment by bboreham** [comment]
> The original description would still be valid I think, 

I say this statement is not true, given Cortex etc. manage to do it:

> multi-tenant applications built on Prometheus [...] Currently, this is only possible by filtering the API response

**Comment by bwplotka** [comment]
(During Prometheus bug scrub meeting)

Just to make sure it's clear: Sounds like maintainers would accept such a change, looks like there are two use cases (workable UI and label-based multi-tenancy).

**Comment by saswatamcode** [comment]
Thanks! Have updated it to:
> Currently, one possible way of doing this is by filtering the API response.

File: web/api/v1/api_test.go
Comment (qinxx108) [reply web/api/v1/api_test.go:215]: remove this parameter as we are not use this parameter at alerts api

File: web/api/v1/errors_test.go
Comment (qinxx108) [reply web/api/v1/errors_test.go:256]: remove this parameter as we are not use this parameter at alerts api

File: web/api/v1/api_test.go
Comment (qinxx108) [reply web/api/v1/api_test.go:2264]: added

File: docs/querying/api.md
Comment (qinxx108) [reply docs/querying/api.md:682]: fixed

**Comment by qinxx108** [comment]
@saswatamcode @roidelapluie @bboreham  I've addressed the comments. Please take another look if you have time!

Ty!

**Comment by yeya24** [comment]
Hi Prometheus maintainers @bboreham @roidelapluie @bwplotka, it looks like there is a concensus to accept this change. It would be great help if any of the maintainer could take a look. Thanks!

**Comment by bwplotka** [comment]
(https://cloud-native.slack.com/archives/C01AUBA4PFE/p1720546903110739) discussion why it's not merged. TL;DR: Comments not adresseed (plus now needs rebase). 

But generally looks good, just let's agree on implementation and defaults.

File: rules/manager.go
Comment (saswatamcode) [reply rules/manager.go]: Yup, I think this is what we are doing here. `matches` ANDs all the matchers in each set i.e return false if any of them don't match, and then `matchesMatcherSet` takes the set of those results and ORs them i.e returns true if any of them are true. Added a comment for this.

**Review by bwplotka — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
